README.md
whoarder
========

**whoarder** converts your Kindle's `My Clippings.txt` file to a more pleasant, sortable, filterable HTML file:
![Whoarder screenshot](whoarder-screenshot.png)

Requirements
------------

- Python 3.3 (so far I only tested with 3.3 on Linux, tests and patches to increase compatibility very welcome)
- The `jinja2` and `chardet2` modules (install them from pip)
- Only tested on a `My Clippings.txt` file produced by a Kindle Paperwhite (ok/ko reports for other devices and test data welcome)
    - Kindle Fire & Kindle Fire HD are not supported, since they do note create the `My Clippings.txt` file. If you know where to dig that data for those versions, patches welcome.

Usage
--------------------

**Command-line**:

Run `whoarder.py /path/to/My Clippings.txt [destination]` . If `destination` is omitted, the output HTML will be written in the same place (overwriting any pre-existing HTML).

**As module**: 

    from clippings import Clippings
    clippings = Clippings(args.source, args.destination)  # contains a 'clippings' dict containing the information
    clippings.export_clippings()  # exports as HTML

**Tests**:

Some unit tests are in `test.py`, along with sample data in `test.txt`.

Similar Software
----------------

- Web services
    - [Clippings Converter](http://www.clippingsconverter.com/)
    - [Clipper](http://www.claybavor.com/clipper/)
- Offline
    - (PHP) [kindle-split-by-book](https://gist.github.com/elvisciotti/1783585)

License and contact
-------------------

Licensed under the BSD license, 2013, [ronan@jouchet.fr](mailto:ronan@jouchet.fr) / [@ronjouch](https://twitter.com/ronjouch)

CHANGES.txt
0.3.0
=====
* Made parser less fragile regarding notes format:
  * Stopped trying to parse author as "lastName, firstName", which isn't always respected
  * Pre-process bookmarks, which do not have "On Page ..." in their second line
  * Ignore case (e.g. on "on", "Page")
  * When a line import fails, print a failure log instead of crashing

0.2.0, 0.2.1, 0.2.2
===================
* Mostly packaging/pip fixes

0.1
===
* Initial release

test.py
from whoarder.clippings import Clippings
import unittest


class TestWrongImport(unittest.TestCase):

    def test_wrong_path(self):
        '''
        converting a non-existent file should return a FileNotFoundError
        '''
        with self.assertRaises(FileNotFoundError):
            self.clippings = Clippings('bar.txt')


class TestImport(unittest.TestCase):

    def setUp(self):
        clippingsimporter = Clippings('test.txt')
        self.clippings = clippingsimporter.clippings

    def test_first_clipping(self):
        self.assertEqual(self.clippings[0]['book'], '<book>')
        self.assertEqual(self.clippings[0]['author_first_name'],
                         '<author_first_name>')
        self.assertEqual(self.clippings[0]['author_last_name'],
                         '<author_last_name>')
        self.assertEqual(self.clippings[0]['type'], '<type>')
        self.assertEqual(self.clippings[0]['page'], 'Page 42')
        self.assertEqual(self.clippings[0]['location'], '123-321')
        self.assertEqual(self.clippings[0]['date'], '<date>')

    def test_count(self):
        '''
        test.txt' should yield 14 clippings (either note or highlight)
        '''
        self.assertEqual(len(self.clippings), 15)

    def test_count_notes(self):
        '''
        test.txt should yield 1 note
        '''
        print(self.clippings)
        notes = [i for i in self.clippings if i['type'] == 'Note']
        self.assertEqual(len(notes), 1)

    def test_count_highlights(self):
        '''
        test.txt should yield 13 highlights
        '''
        highlights = [i for i in self.clippings if i['type'] == 'Highlight']
        self.assertEqual(len(highlights), 13)

    def test_bom_stripped(self):
        '''
        Ensure repeated BOMs incorrectly written by Kindle are stripped
        '''
        for clipping in self.clippings:
            book_first_char = clipping['book'][0]
            self.assertNotEqual(book_first_char, "\ufeff")

    def test_presence_book(self):
        '''
        Each clipping should reference a book.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['book'])

    def test_presence_author_first_name(self):
        '''
        Each clipping should reference its author's first name.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['author_first_name'])

    def test_presence_author_last_name(self):
        '''
        Each clipping should reference its author's last name.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['author_last_name'])

    def test_presence_location(self):
        '''
        Each clipping should reference the kindle location it appeared on.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['location'])

    def test_presence_date(self):
        '''
        Each clipping should have a date.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['date'])

    def test_presence_contents(self):
        '''
        Each clipping should have contents.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['contents'])

#     def test_presence_page(self):
#         '''
#                 # each clipping should reference the page it appeared on
#         '''
#         for clipping in self.clippings:
#             self.assertIsNotNone(clipping['page'])

if __name__ == '__main__':
    unittest.main()


whoarder/test/__init__.py
from whoarder.clippings import Clippings
import unittest


class TestImport(unittest.TestCase):

    def setUp(self):
        clippingsimporter = Clippings('test.txt')
        self.clippings = clippingsimporter.clippings

    def test_first_clipping(self):
        self.assertEqual(self.clippings[0]['book'], '_BOOK_')
        self.assertEqual(self.clippings[0]['author'], '_AUTHOR_LAST_NAME_, _AUTHOR_FIRST_NAME_')
        self.assertEqual(self.clippings[0]['type'], '_TYPE_')
        self.assertEqual(self.clippings[0]['page'], 'Page 42')
        self.assertEqual(self.clippings[0]['location'], '123-321')
        self.assertEqual(self.clippings[0]['date'], '_DATE_')

    def test_count(self):
        '''
        test.txt should yield a certain number of clippings.
        '''
        self.assertEqual(len(self.clippings), 17)

    def test_count_notes(self):
        '''
        test.txt should yield 1 note
        '''
        print(self.clippings)
        notes = [i for i in self.clippings if i['type'] == 'Note']
        self.assertEqual(len(notes), 1)

    def test_count_highlights(self):
        '''
        test.txt should yield 13 highlights
        '''
        highlights = [i for i in self.clippings if i['type'] == 'Highlight']
        self.assertEqual(len(highlights), 13)

    def test_count_bookmarks(self):
        '''
        test.txt should yield 2 bookmarks
        '''
        bookmarks = [i for i in self.clippings if i['type'] == 'Bookmark']
        self.assertEqual(len(bookmarks), 2)

    def test_bom_stripped(self):
        '''
        Ensure repeated BOMs incorrectly written by Kindle are stripped
        '''
        for clipping in self.clippings:
            book_first_char = clipping['book'][0]
            self.assertNotEqual(book_first_char, "\ufeff")

    def test_presence_book(self):
        '''
        Each clipping should reference a book.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['book'])

    def test_presence_author(self):
        '''
        Each clipping should reference its author's name.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['author'])

    def test_presence_location(self):
        '''
        Each clipping should reference the kindle location it appeared on.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['location'])

    def test_presence_date(self):
        '''
        Each clipping should have a date.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['date'])

    def test_presence_contents(self):
        '''
        Each clipping should have contents.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['contents'])


class TestWrongImport(unittest.TestCase):

    def test_wrong_path(self):
        '''
        converting a non-existent file should return a FileNotFoundError
        '''
        with self.assertRaises(FileNotFoundError):
            self.clippings = Clippings('bar.txt')

if __name__ == '__main__':
    unittest.main()


whoarder/tests/__init__.py
from whoarder.clippings import Clippings
import unittest


class TestImport(unittest.TestCase):

    def setUp(self):
        clippingsimporter = Clippings('test.txt')
        self.clippings = clippingsimporter.clippings

    def test_first_clipping(self):
        self.assertEqual(self.clippings[0]['book'], '<book>')
        self.assertEqual(self.clippings[0]['author_first_name'],
                         '<author_first_name>')
        self.assertEqual(self.clippings[0]['author_last_name'],
                         '<author_last_name>')
        self.assertEqual(self.clippings[0]['type'], '<type>')
        self.assertEqual(self.clippings[0]['page'], 'Page 42')
        self.assertEqual(self.clippings[0]['location'], '123-321')
        self.assertEqual(self.clippings[0]['date'], '<date>')

    def test_count(self):
        '''
        test.txt' should yield 14 clippings (either note or highlight)
        '''
        self.assertEqual(len(self.clippings), 15)

    def test_count_notes(self):
        '''
        test.txt should yield 1 note
        '''
        print(self.clippings)
        notes = [i for i in self.clippings if i['type'] == 'Note']
        self.assertEqual(len(notes), 1)

    def test_count_highlights(self):
        '''
        test.txt should yield 13 highlights
        '''
        highlights = [i for i in self.clippings if i['type'] == 'Highlight']
        self.assertEqual(len(highlights), 13)

    def test_bom_stripped(self):
        '''
        Ensure repeated BOMs incorrectly written by Kindle are stripped
        '''
        for clipping in self.clippings:
            book_first_char = clipping['book'][0]
            self.assertNotEqual(book_first_char, "\ufeff")

    def test_presence_book(self):
        '''
        Each clipping should reference a book.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['book'])

    def test_presence_author_first_name(self):
        '''
        Each clipping should reference its author's first name.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['author_first_name'])

    def test_presence_author_last_name(self):
        '''
        Each clipping should reference its author's last name.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['author_last_name'])

    def test_presence_location(self):
        '''
        Each clipping should reference the kindle location it appeared on.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['location'])

    def test_presence_date(self):
        '''
        Each clipping should have a date.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['date'])

    def test_presence_contents(self):
        '''
        Each clipping should have contents.
        '''
        for clipping in self.clippings:
            self.assertIsNotNone(clipping['contents'])


class TestWrongImport(unittest.TestCase):

    def test_wrong_path(self):
        '''
        converting a non-existent file should return a FileNotFoundError
        '''
        with self.assertRaises(FileNotFoundError):
            self.clippings = Clippings('bar.txt')

if __name__ == '__main__':
    unittest.main()


whoarder/whoarder.py
#!/usr/bin/env python3

import argparse

if __name__ == '__main__':

    # Boilerplate to allow running as script directly. Avoids error below:
    # SystemError: Parent module '' not loaded, cannot perform relative import
    # See http://stackoverflow.com/questions/2943847/
    if __package__ is None:
        import sys
        import os
        abspath = os.path.abspath(__file__)
        parent_dir = os.path.dirname(os.path.dirname(abspath))
        sys.path.insert(0, parent_dir)
        from whoarder.clippings import Clippings
        del sys, os

    parser = argparse.ArgumentParser(description="whoarder converts Kindle \
                        'My Clippings.txt' files to more pleasant HTML.")
    parser.add_argument('source',
                        help='Path to the source file, stored by Kindle in \
                        /Media/Kindle/documents/My Clippings.txt.')
    parser.add_argument('destination',
                        help='Target HTML file. If omitted, a .html bearing \
                        the same name as the input .txt file will be used.',
                        nargs='?', default=None)
    args = parser.parse_args()

    clippings = Clippings(args.source, args.destination)
    clippings.export_clippings()
    print('Successfully wrote ' + clippings.dest + "\n")


whoarder/clippings.py
import chardet
import codecs
import os
import re


class Clippings(object):

    def __init__(self, source, dest=None):
        '''
        Prepares the import and store it into the 'clippings' dict.
        '''
        self.source = source
        self.dest = self._get_default_dest() if dest is None else dest
        self.book_author_couples = ()
        self.clippings = []
        self._fetch()

    def _get_default_dest(self):
        '''
        When no destination is specified, output to <InputFilename>.html
        '''
        source_full_path = os.path.realpath(self.source)
        dirname, filename_with_ext = os.path.split(source_full_path)
        filename = os.path.splitext(filename_with_ext)[0]
        default_destination = os.path.join(dirname, filename + '.html')
        return default_destination

    def _fetch(self):
        '''
        Imports clippings and book_author_couples from the source file
        '''
        clippings = ClippingsIterator(self.source)
        for clipping in clippings:
            self.clippings.append(clipping)

        # will be useful in the HTML to group by book/author
        self.book_author_couples = set((clipping['book'], clipping['author'])
                                       for clipping in self.clippings)

    def export_to_html(self):
        '''
        Output the clippings dict to HTML, using a Jinja2 template
        '''
        from jinja2 import Environment, PackageLoader  # available from pip
        env = Environment(loader=PackageLoader('whoarder', 'templates'),
                          autoescape=True,
                          extensions=['jinja2.ext.autoescape'])
        template = env.get_template('template1.html')
        render = template.render(clippings=self.clippings,
                                 book_author_couples=self.book_author_couples)

        with open(self.dest, mode='w', encoding='utf-8') as output:
            output.write(render)


class ClippingsIterator(object):
    '''
    Iterator that abstracts the Kindle format and spits a dict per clipping.
    A 'clipping' can be either a Highlight or a Note, and is (as far as I
    know, on my Kindle) a succession of five lines (see ex. and regexes below):
    - Lines 1 & 2 contain metadata
    - Line 3 is empty
    - Line 4 is the clipping
    - Line 5 is the separator

    Example:
    <book> (<author_last_name>, <author_first_name>)
    - Your <type> on Page <page> | Location <locs>-<loce> | Added on <date>

    <contents>
    ==========
    '''

    _clipping_separator = '==========\n'
    _clipping_line1 = re.compile(r'''
        ^(?P<book>.*)                            # Le Petit Prince
        \ \((?P<author>.*)\)$                    #  (De Saint-Exupery, Antoine)
        ''', re.VERBOSE | re.IGNORECASE)
    _clipping_line2 = re.compile(r'''
        ^-\ Your\ (?P<type>\w*)                         # Your Highlight
        \ (?:on\ )?(?P<page>Unnumbered\ Page|Page\ .*)  #  on Page 42
        \ \|\ (?:on\ )?Location\ (?P<location>.*)       #  | Location 123-321
        \ \|\ Added\ on\ (?P<date>.*)$                  #  | Added on...
        ''', re.VERBOSE | re.IGNORECASE)

    def __init__(self, source):
        detected_encoding = _detect_encoding(source)
        self.source_file = open(source, mode='r', encoding=detected_encoding)

    def __iter__(self):
        return self

    def __next__(self):
        clipping_buffer = []
        count = 1
        while True:
            if count > 5:
                raise InvalidFormatException('''Input file doesn't seem to be
                a clippings file, separators are missing or damaged''')
            if self.source_file.closed:
                raise StopIteration

            line = self.source_file.readline()

            if not line:
                self.source_file.close()
                raise StopIteration
            elif line != self._clipping_separator:
                # Kindle writes a FEFF BOM at the start of each clipping
                # (i.e. every 6 lines), which is clearly wrong. We strip it.
                if line[0] == "\ufeff":
                    line = line.replace("\ufeff", "")
                clipping_buffer.append(line.strip())
                count += 1
            else:
                break

        if 'Page' not in clipping_buffer[1] and 'page' not in clipping_buffer[1]:
            clipping_buffer[1] = re.sub(r'(- Your .*?) (.*)',
                                        r'\1 on Unnumbered Page | \2',
                                        clipping_buffer[1])

        try:
            line_dict = self._clipping_line1.search(clipping_buffer[0]).groupdict()
            line_dic2 = self._clipping_line2.search(clipping_buffer[1]).groupdict()
            line_dict.update(line_dic2)
            line_dict['contents'] = clipping_buffer[3]
            return line_dict
        except AttributeError:
            print("Failed to import the following note, please report to https://github.com/ronjouch/whoarder :\n  {0}\n".format(clipping_buffer))
            return self.__next__()


def _detect_encoding(source):
    '''
    Returns the encoding of the source file, using chardet.
    '''
    rawdata = open(source, "rb").read()
    # chardet detects UTF-8 with BOM as 'UTF-8' (I don't know why), i.e.
    # fails to notify us about the BOM, resulting in a string prepended
    # with \ufeff, so we manually detect and set the utf-8-sig encoding
    if rawdata.startswith(codecs.BOM_UTF8):
        detected_encoding = 'utf-8-sig'
    else:
        result = chardet.detect(rawdata)
        detected_encoding = result['encoding']
    return detected_encoding


class InvalidFormatException(Exception):
    pass
